enhancements/20230421-adjust-the-scale-of-resource-quota-during-migration.md

```diff
@@ -0,0 +1,219 @@
+
+In addition, validators are added to the VM and Restore CRs to ensure that there are sufficient resources to create, start, or restore a VM.
+
+When a migration is created, the Migration Controller adds an annotation to the ResourceQuota in its namespace, as follows:
+
+```yaml
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+  annotations:
+    harvesterhci.io/migratingVMs: '{"vm-01":{"limits.cpu":"1","limits.memory":"1267028Ki"}}'
+  labels:
+    cattle.io/creator: norman
+    resourcequota.management.cattle.io/default-resource-quota: "true"
+  name: default-x7fpr
+  namespace: rs
+spec:
+  hard:
+    limits.cpu: "4"
+    limits.memory: 4952Mi
+status:
+  hard:
+    limits.cpu: "4"
+    limits.memory: 4952Mi
+  used:
+    limits.cpu: "4"
+    limits.memory: 5068112Ki
+```
+
+The key is `vm_name` with a value of the corresponding VM's Pod resources limits. When Migration is in Pending, Harvester should automatically scale up the ResourceQuota to meet the resource requirements for the VM migration. Once the VM migration is complete, Harvester should automatically scale down and delete the key.
+
+Due to the unreliability of the Webhook, the VM Controller will perform a second verification for VMs. If the resources are insufficient, the VM's RunStrategy will be changed to Halted, and an event will be sent indicating the lack of resources.
+
+**Upgrade**
+
+The implementation is similar to the above. If a large number of VM migrations occur in a node, causing the ResourceQuota to scale up and Migration to be in Pending, the VMs that are already started may race for resources. To avoid this situation, the actual available resources for the VMs are calculated using the formula: actual limit - (used - migrations), and the result is compared to see if there are enough resources for the VM to start.
+
+**Maintenance**
+
+Similar to Upgrade, but if there are single replica VMs in the node, the system will shut them down, and they will not be counted toward used resources, so no processing is needed.
+
+**Changing ResourceQuota**
+
+When a user changes ResourceQuota, the system checks whether there are any migrating VMs in the annotations. If there are, the user will be prompted that they cannot change the ResourceQuota while the VM is being migrated.
+
+**Overhead**
+
+Before creating a Pod instance for each VM, the upstream first calculates the Overhead resources of the VM and adds them to the VM Pod before deployment. During VM migration, the target Pod already contains the Overhead resources and the resource specifications of that Pod are obtained during migration. In addition, the Overhead is calculated before the VM is started([calculation](https://github.com/kubevirt/kubevirt/blob/2bb88c3d35d33177ea16c0f1e9fffdef1fd350c6/pkg/virt-controller/services/template.go#L1804-L1893)) And overlay it with the VM resource specifications and verify that it can be started.
```

Comment:
```suggestion
Before creating a pod instance for each VM, the upstream first calculates the Overhead resources of the VM and adds them to the VM pod before deployment. During VM migration, the target pod already contains the Overhead resources, and that pod's resource specifications are obtained during migration. In addition, the Overhead resources are calculated before the VM starts and then overlayed with the VM resource specifications to verify that it can be started. For details, refer to [GetMemoryOverhead](https://github.com/kubevirt/kubevirt/blob/2bb88c3d35d33177ea16c0f1e9fffdef1fd350c6/pkg/virt-controller/services/template.go#L1804-L1893).
```